src/CacheControl/CacheControlServiceProvider.php

```diff
@@ -0,0 +1,98 @@
+            RegisterDirectiveNamespaces::class,
+            static function (): string {
+                return __NAMESPACE__;
+            }
+        );
+
+        $dispatcher->listen(
+            StartExecution::class,
+            function (StartExecution $StartExecution) use ($cacheControl) {
+                $typeInfo = new TypeInfo($StartExecution->schema);
+                Visitor::visit($StartExecution->query, Visitor::visitWithTypeInfo($typeInfo, [
+                    NodeKind::FIELD => function (FieldNode $node) use ($typeInfo, $cacheControl): void {
+                        $field = $typeInfo->getFieldDef();
+                        // @phpstan-ignore-next-line can be null, remove ignore with graphql-php 15
+                        if (null === $field) {
+                            return;
+                        }
+
+                        $nodeType = $field->getType();
+                        if ($nodeType instanceof NonNull || $nodeType instanceof ListOfType) {
+                            do {
+                                $nodeType = $nodeType->getOfType();
+                            } while ($nodeType instanceof NonNull || $nodeType instanceof ListOfType);
+                        }
+
+                        if (! $nodeType instanceof ScalarType) {
+                            $maxAge = 0;
+                        }
+
+                        if (isset($field->astNode->directives)) {
+                            $cacheControlDirective = new Collection($field->astNode->directives);
+                            $cacheControlDirective = $cacheControlDirective->where('name.value', 'cacheControl')->first();
+                            if (! is_null($cacheControlDirective) && key_exists('arguments', $cacheControlDirective->toArray())) {
+                                $arguments = new Collection($cacheControlDirective->arguments);
+                                $maxAge
+                                    = $arguments->where('name.value', 'maxAge')->pluck('value.value')->first()
+                                    ?? 0;
+                                $scope
+                                    = $arguments->where('name.value', 'scope')->pluck('value.value')->first()
+                                    ?? 'PUBLIC';
+                            }
+                        }
+
+                        if (isset($maxAge)) {
+                            $cacheControl->addToMaxAgeList($maxAge);
+                        }
+                        if (isset($scope)) {
+                            $cacheControl->addToScopeList($scope);
+                        }
+                    },
+                ]));
+            }
+        );
+
+        $dispatcher->listen(
+            EndRequest::class,
+            function (EndRequest $request) use ($cacheControl): void {
+                $maxAge = $cacheControl->calculateMaxAge();
+                if (0 != $maxAge) {
+                    $request->response->setMaxAge($maxAge);
```

Comment:
We can extract both `$response` and `$headers` to avoid repeated property access.